src/Products/Viewer/Cache/CrossProcessLock.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Security.AccessControl;
+using System.Security.Principal;
+using System.Threading;
+
+namespace GroupDocs.Viewer.MVC.Products.Viewer.Cache
+{
+    sealed class CrossProcessLock : IDisposable
+    {
+        private const int MaxMutexIdLength = 260;
+        private readonly Mutex mutex;
+        private readonly bool isMutexOwner;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CrossProcessLock"/> class.
+        /// </summary>
+        /// <param name="path">The file path.</param>
+        public CrossProcessLock(string path)
+        {
+            string mutexId = this.GetMutexId(path);
+            MutexSecurity mutexSecurity = this.GetMutexSecurity();
+            try
+            {
+                bool createdNew;
+                this.mutex = new Mutex(false, mutexId, out createdNew, mutexSecurity);
+
+                try
+                {
+                    this.isMutexOwner = this.mutex.WaitOne(10 * 60 * 1000);
+                }
+                catch (AbandonedMutexException)
+                {
+                    // NOTE: Log the fact the mutex was abandoned in another process, it will still get acquired
+                }
+            }
+            catch (WaitHandleCannotBeOpenedException ex)
+            {
+                // the mutex cannot be opened, probably because a Win32 object of a different
+                // type with the same name already exists.
+                throw new Exception("Mutex can't be opened: " + ex.Message);
+            }
+            catch (UnauthorizedAccessException ex)
+            {
+                // the mutex exists, but the current process or thread token does not
+                // have permission to open the mutex with SYNCHRONIZE | MUTEX_MODIFY rights.
+                throw new Exception("Current process does not have permission to open the mutex: " + ex.Message);
```

Comment:
Codacy found an issue: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5933705)